x-pack/metricbeat/module/aws/sqs/sqs.go

```diff
@@ -0,0 +1,194 @@
+// any MetricSet specific configuration options if there are any.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	logger := logp.NewLogger(aws.ModuleName)
+
+	moduleConfig := aws.Config{}
+	if err := base.Module().UnpackConfig(&moduleConfig); err != nil {
+		return nil, err
+	}
+
+	if moduleConfig.Period == "" {
+		err := errors.New("period is not set in AWS module config")
+		logger.Error(err)
+	}
+
+	metricSet, err := aws.NewMetricSet(base)
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating aws metricset")
+	}
+
+	// Check if period is set to be multiple of 300s
+	remainder := metricSet.PeriodInSec % 300
+	if remainder != 0 {
+		err := errors.New("period needs to be set to 300s (or a multiple of 300s). " +
+			"To avoid data missing or extra costs, please make sure period is set correctly in config.yml")
+		logger.Info(err)
+	}
+
+	return &MetricSet{
+		MetricSet: metricSet,
+		logger:    logger,
+	}, nil
+}
+
+// Fetch methods implements the data gathering and data conversion to the right
+// format. It publishes the event which is then forwarded to the output. In case
+// of an error set the Error field of mb.Event or simply call report.Error().
+func (m *MetricSet) Fetch(report mb.ReporterV2) {
+	namespace := "AWS/SQS"
+	// Get startTime and endTime
+	startTime, endTime, err := aws.GetStartTimeEndTime(m.DurationString)
+	if err != nil {
+		logp.Error(errors.Wrap(err, "Error ParseDuration"))
+		m.logger.Error(err.Error())
+		report.Error(err)
+		return
+	}
+
+	for _, regionName := range m.MetricSet.RegionsList {
+		m.MetricSet.AwsConfig.Region = regionName
+		svcCloudwatch := cloudwatch.New(*m.MetricSet.AwsConfig)
+		listMetricsOutput, err := aws.GetListMetricsOutput(namespace, regionName, svcCloudwatch)
+		if err != nil {
+			m.logger.Error(err.Error())
+			report.Error(err)
+			continue
+		}
+
+		if listMetricsOutput == nil || len(listMetricsOutput) == 0 {
+			continue
+		}
+
+		metricDataQueries := constructMetricQueries(listMetricsOutput, int64(m.PeriodInSec))
+		if len(metricDataQueries) == 0 {
+			continue
+		}
+
+		init := true
+		getMetricDataOutput := &cloudwatch.GetMetricDataOutput{NextToken: nil}
+		for init || getMetricDataOutput.NextToken != nil {
+			init = false
+			output, err := aws.GetMetricDataPerRegion(metricDataQueries, getMetricDataOutput.NextToken, svcCloudwatch, startTime, endTime)
+			if err != nil {
+				err = errors.Wrap(err, "getMetricDataPerRegion failed, skipping region "+regionName)
+				m.logger.Error(err.Error())
+				report.Error(err)
+				continue
+			}
+			getMetricDataOutput.MetricDataResults = append(getMetricDataOutput.MetricDataResults, output.MetricDataResults...)
+		}
+
+		// Create Cloudwatch Events for SQS
+		event, err := createSQSEvents(getMetricDataOutput.MetricDataResults, metricsetName, regionName, schemaRequestFields)
+		if err != nil {
+			m.logger.Error(err.Error())
+			event.Error = err
+			report.Event(event)
```

Comment:
Why not use report.Error(...) here?

Reply:
I'm not sure... report.Error is what I want here... Thanks!

Reply:
I think it is correct as is if you want to report an event with whatever data it has AND an error.
In this case it seems that this would be an error with some metadata (`service.name`, `cloud.region`) and without any metrics, I think this could be acceptable.
If this is the case and this is the only thing blocking this PR I think this would be ready to go.